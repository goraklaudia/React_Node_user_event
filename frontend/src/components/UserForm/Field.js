import { TextField } from "@mui/material";

export const Field = ({id, label, type, value, setValue, error}) => {
    return (
            <TextField
                id={id}
                label={label}
                type={type}
                value={value}
                margin="dense"
                fullWidth
                error={error.value}
                helperText={error.text}
                onChange={(e)=> {setValue(e.target.value)}}
                InputLabelProps={{ shrink: true }}
            />
    );
};