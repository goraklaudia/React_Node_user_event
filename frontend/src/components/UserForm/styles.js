import { makeStyles } from '@mui/styles';

const useStyles = makeStyles({
    mainBox: {
        margin: '10% 20%',
        width: '60%',
        justifyContent: 'center',
        alignItems: 'center',
        border: '1px solid gray',
        borderRadius: '5px',
        textAlign: 'center',
        // [theme.breakpoints.down('xs')]: {
        //     width: '90%',
        //     margin: '10% 5%',
        // }
    },
    boxField: {
        margin: '5% 5%',
        width: '96%',
    }
});
export default useStyles;