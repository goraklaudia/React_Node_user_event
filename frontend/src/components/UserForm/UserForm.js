import { useState } from "react";
import { Typography, Button } from "@mui/material";
import { useStyles } from "./styles";
import { Field } from "./Field";
import { validateDate, validateEmail } from "../../utils/validations";
import { fetchHandler } from '../../utils/fetchHandler';
import Alert from '../Alert/Alert';

const UserForm = () => {
    const classes = useStyles();

    const [firstName, setFirstName] = useState("");
    const [lastName, setLastName] = useState("");
    const [email, setEmail] = useState("");
    const [date, setDate] = useState("");

    const [errorFirstName, setErrorFirstName] = useState({});
    const [errorLastName, setErrorLastName] = useState({});
    const [errorEmail, setErrorEmail] = useState({});
    const [errorDate, setErrorDate] = useState({});

    const [alert, setAlert] = useState({})

    const generateArrayOfFields = () => {
        return [
            {
                id: "firstName",
                label: "First name *",
                type: "text",
                value: firstName,
                setValue: setFirstName,
                error: errorFirstName
            },
            {
                id: "lastName",
                label: "Last name *",
                type: "text",
                value: lastName,
                setValue: setLastName,
                error: errorLastName
            },
            {
                id: "email",
                label: "Email *",
                type: "email",
                value: email,
                setValue: setEmail,
                error: errorEmail
            },
            {
                id: "date",
                label: "Date *",
                type: "date",
                value: date,
                setValue: setDate,
                error: errorDate
            },
        ];
    };

    const validateFields = () => {
        let allFieldsValid = true;

        if (!date || !validateDate(date)) {
            setErrorDate({ value: true, text: 'Date is invalid' });
            allFieldsValid = false;
        } else {
            setErrorDate({ value: false, text: '' })
        }

        if (!email || !validateEmail(email)) {
            setErrorEmail({ value: true, text: 'Email is invalid' });
            allFieldsValid = false;
        } else {
            setErrorEmail({ value: false, text: '' });
        }

        if (!firstName) {
            setErrorFirstName({ value: true, text: 'First name is invalid' })
            allFieldsValid = false;
        } else {
            setErrorFirstName({ value: false, text: '' });
        }

        if (!lastName) {
            setErrorLastName({ value: true, text: 'Last name is invalid' });
            allFieldsValid = false;
        } else {
            setErrorLastName({ value: false, text: '' });
        }

        return allFieldsValid;
    };

    const handleSave = () => {
        const validField = validateFields();

        if (validField) {
            let options = {
                method: 'POST',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    firstName,
                    lastName,
                    email,
                    date
                })
            }
            fetchHandler('/user', options, setAlert);
        }
    };

    return (
        <div className={classes.mainBox}>
            <Alert msg={alert.msg} open={alert.open} setOpen={setAlert}/>
            <Typography variant="h4" className={classes.title}>
                User Form
            </Typography>
            {generateArrayOfFields().map(({ id, label, type, value, setValue, error }) => {
                return (
                    <Field
                        key={id}
                        id={id}
                        label={label}
                        type={type}
                        value={value}
                        setValue={setValue}
                        error={error}
                    />
                );
            })}
            <Button
                variant="contained"
                size="large"
                onClick={handleSave}
            >
                Save
            </Button>
        </div>
    );
};

export default UserForm;
