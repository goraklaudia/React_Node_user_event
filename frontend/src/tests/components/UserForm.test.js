import React from 'react';
import { create } from 'react-test-renderer';
import UserForm from '../../components/UserForm/UserForm'

describe('Component UserForm', () => {
    it('should render correctly', () => {
        const component = create(<UserForm/>);
        expect(component.toJSON()).toMatchSnapshot();
    });
})