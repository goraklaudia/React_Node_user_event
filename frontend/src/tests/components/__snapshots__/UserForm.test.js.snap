// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component UserForm should render correctly 1`] = `
<div
  className="makeStyles-mainBox-1"
>
  <h4
    className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
  >
    User Form
  </h4>
  <div
    className="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      htmlFor="firstName"
      id="firstName-label"
    >
      First name *
    </label>
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id="firstName"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-186xcr5"
        >
          <span>
            First name *
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      htmlFor="lastName"
      id="lastName-label"
    >
      Last name *
    </label>
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id="lastName"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-186xcr5"
        >
          <span>
            Last name *
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      htmlFor="email"
      id="email-label"
    >
      Email *
    </label>
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id="email"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="email"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-186xcr5"
        >
          <span>
            Email *
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      htmlFor="date"
      id="date-label"
    >
      Date *
    </label>
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id="date"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="date"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-186xcr5"
        >
          <span>
            Date *
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <button
    className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root css-jh47zj-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Save
  </button>
</div>
`;
